File: lib/core_services/comic/ab_comic_service.dart
import 'dart:typed_data';

import 'main.dart';

export 'package:hoyomi/core_services/service.dart';

abstract class ABComicService extends Service
    with ComicFollowGeneralMixin, ComicWatchPageGeneralMixin {
  Future<Uint8List> fetchPage(Uint8List buffer, OImage source) {
    throw UnimplementedError();
  }

  Future<ComicCategory> getCategory({
    required String categoryId,
    required int page,
    required Map<String, List<String>?> filters,
  });

  Future<ComicCategory> getExplorer({
    required int page,
    required Map<String, List<String>?> filters,
  }) {
    throw UnimplementedError();
  }

  // Utils
  ComicModes getComicModes(MetaComic comic) {
    throw UnimplementedError();
  }

  Future<MetaComic> getDetails(String comicId);
  Future<List<OImage>> getPages(String manga, String chap);
  Future<List<Comic>> getSuggest(MetaComic comic, {int? page}) {
    throw UnimplementedError();
  }

  Future<String> getURL(String comicId, {String? chapterId});

  Future<ComicHome> home();

  Future<ComicCategory> search({
    required String keyword,
    required int page,
    required Map<String, List<String>?> filters,
    required bool quick,
  });
}

File: lib/core_services/comic/ab_comic_service.ts
import { BaseService } from "../service"
import {
  type ComicHome,
  type ComicCategory,
  type MetaComic,
  type OImage,
  type Comic,
  ComicModes
} from "../main"

export abstract class ABComicService extends BaseService {
  readonly type = "comic"

  abstract getURL(comicId: string, chapterId?: string): Promise<string>

  abstract home(): Promise<ComicHome>

  abstract getCategory(params: {
    categoryId: string
    page: number
    filters: { [key: string]: string[] | null }
  }): Promise<ComicCategory>

  getExplorer(
    page: number,
    filters?: { [key: string]: string[] | null }
  ): Promise<ComicCategory> {
    throw new UnimplementedError()
  }

  abstract getDetails(comicId: string): Promise<MetaComic>
  abstract getPages(manga: string, chap: string): Promise<OImage[]>

  fetchPage(buffer: Uint8Array, source: OImage): Promise<Uint8Array> {
    throw new UnimplementedError()
  }

  getSuggest(comic: MetaComic, page?: number): Promise<Comic[]> {
    throw new UnimplementedError()
  }

  abstract search(params: {
    keyword: string
    page: number
    filters: { [key: string]: string[] | null }
    quick: boolean
  }): Promise<ComicCategory>

  // Utils
  getComicModes(comic: MetaComic): ComicModes {
    throw new UnimplementedError()
  }
}
